xonsh/tools.py

```diff
@@ -1634,13 +1654,45 @@ def lazyget(self, key, default=None):
         """A lazy value getter."""
         return self._cmds_cache.get(key, default)
 
+    def locate_binary(self, name):
+        """Locates an executable on the file system using the cache."""
+        possibilities = self.get_possible_names(name)
+
+        if ON_WINDOWS:
+            from xonsh.dirstack import _get_cwd  # circular dependency
+            # Windows users expect to be able to execute files in the same
+            # directory without `./`
+            cwd = _get_cwd()
+            local_bin = next((
+                full_name for full_name in possibilities
+                if os.path.isfile(full_name)
+            ), None)
+            if local_bin:
+                return os.path.abspath(os.path.relpath(local_bin, cwd))
+
+        cmds = self.all_commands
```

Comment:
Basically, you only want to access `all_commands` property ones. Afterwards use the lazy methods.